### PR TITLE
[MM-45008]: adjusted behavior of formattingbar in advanced text editor

### PR DESCRIPTION
**Pull request by michelengelen**
#### Summary
adjustments to the behavior of the formatting bar toggle functionality in the Advanced Text Editor.

Specific changes:
1. formatting bar now slides up/down when toggled
2. background color removed
3. spacing between icons reduced
4. active state color change on the toggle removed
5. chevron active indicator removed from toggle
6. tooltip changes from `Show Formatting` to `Hide Formatting` when toggled
7. moved formattingbar toggle, fileupload and emojipicker to the right (next to send button)
8. hides preview button when formatting bar is inactive

#### Ticket Link
[MM-45008](https://mattermost.atlassian.net/browse/MM-45008)

#### Related Pull Requests
n/a

#### Screenshots
n/a

#### Release Note
```release-note
NONE
```

### CODE REVIEW

**Comment by AshishDhama** [comment]
@michelengelen There is some overlap happening, is this the known issue you mentioned about in the thread on Community 
Screenshot 2022-08-30 at 11 57 06 AM

**Comment by michelengelen** [comment]
> @michelengelen There is some overlap happening, is this the known issue you mentioned about in the thread on Community Screenshot 2022-08-30 at 11 57 06 AM

@AshishDhama exactly ... this is already thought on by UX

File: package-lock.json
Comment (M-ZubairAhmed) [inline package-lock.json:8]: please revert this

File: sass/components/_post.scss
Comment (koox00) [inline sass/components/_post.scss:2247]: does this do anything? AFAIK `::after` cannot be used in input elements

Comment (michelengelen) [reply sass/components/_post.scss:2247]: lol ... you are right ... this is a remnant of something I tried ... will remove it

**Review by AshishDhama — APPROVED** [review]

**Comment by jgilliam17** [comment]
Thanks @michelengelen 
Seeing few issues:
1. I am not finding "emoji button & attachment button" when formatting bar is toggled off (per figma and ticket)
2. "Send button" being outside of the textbox seems a bit off. Button is not on the same level with the textbox. And it also makes the rhs look slanted, out of line.
Screen Shot 2022-08-31 at 3 32 56 PM

3. Multiple draft attachments expand the textbox. 
On RHS: multi attachments add a scroll bar and expand the whole pane
On main channel view: the textbox expands with send and preview buttons out of view and reach, no scroll option

https://user-images.githubusercontent.com/52937121/187767323-d6cbe07c-b867-455e-a1c7-20c66a9dc311.mov

4. Preview of long character string (one without breaks/spaces) also expands the textbox. This seems to be very similar to issue 3. above.
Screen Shot 2022-08-31 at 3 47 55 PM

**Comment by michelengelen** [comment]
sry @jgilliam17 ... this is not ready for review and I should have removed the review request from you ... we still need to figure out how to handle the overflowing issue in the best way ... I will mark this PR as a WIP in the meanwhile

**Comment by michelengelen** [comment]
@jgilliam17 @matthewbirtch @andrewbrown00 this is ready for another check! 🎉

**Comment by michelengelen** [comment]
@asaadmahmood feel free to check this out as well ... it leans to what we were trying to do, but with some tweaks here and there

**Review by matthewbirtch — CHANGES_REQUESTED** [review]
Looking really great @michelengelen. 

1. When the toolbar is showing and the preview icon is showing, the text can overlap with it. I think we'll need to add some padding on the right side of the textarea when the preview icon is showing.

image

2. Related to the above, it looks like we add 90px padding when the container scrolls, which seems like maybe too much? Seems like we can reduce to 48 or 56?

image

4. I wonder if we can reduce the bottom padding on the textarea because it seems we can tighten up the space now. Since we no longer have the background color in the toolbar, we can likely tighten this up. It looks like we have it set to 12px bottom padding, and I think we can reduce to 4px.

Let me know what you think

**Comment by michelengelen** [comment]
@matthewbirtch the latest push fixes all your findings.

And additionally I changed the behavior on how the preview button works.
I noticed that when you have preview enabled on a single line and then hide the formatting controls it will leave you in a state where you cannot exit it without using the hotkeys.

I changed it that when the preview is enabled the spacer is still being rendered.

There are other options though:
1. disable preview mode when you hide the formatting controls
2. move the preview into the editor actions section below

Thinking about it option 2 might be the even better approach now that using more space for that section is not a problem anymore. 

Let me know what you think.

**Comment by matthewbirtch** [comment]
@michelengelen what if we just disabled formatting bar toggle button while in preview mode so you couldn't hide the toolbar/preview button while you're previewing?

**Comment by andrewbrown00** [comment]
@michelengelen you can remove me as a reviewer so I don't block this PR

**Comment by michelengelen** [comment]
> @michelengelen what if we just disabled formatting bar toggle button while in preview mode so you couldn't hide the toolbar/preview button while you're previewing?

we could also do that ... let me make that change real quick, so we can get this over the finishing line! >D

**Comment by matthewbirtch** [comment]
That works for me @michelengelen! One last thing I noticed:

- on smaller screens, I'm seeing a doubled-up divider. While we're at it, we should reduce the opacity on the dividers to 16%
image

**Comment by michelengelen** [comment]
@matthewbirtch sry for letting you wait this long. 

I fixed the last findings and reduced the opacity. Could you re-check if that is ok now?

Also @jgilliam17, @koox00 and @M-ZubairAhmed this is ready for another round!

**Comment by matthewbirtch** [comment]
One other thing I just noticed, when the RHS is 400 500px wide, maybe we can just hide dividers - except for the one that shows to the right of the toolbar control....like this:

image

**Comment by esethna** [comment]
@michelengelen checking in, what's next steps on merging this PR? Anything blocking? 

cc// @Willyfrog

**Comment by michelengelen** [comment]
basically only the reviews from @matthewbirtch and @jgilliam17 ... If devs want they can do another round as well

cc// @koox00 @M-ZubairAhmed

File: components/autosize_textarea.tsx
Comment (M-ZubairAhmed) [inline components/autosize_textarea.tsx:43]: i know this was same as before. But should we call this calculation on any update?

**Review by M-ZubairAhmed — APPROVED** [review]
Looks good to me.

File: components/autosize_textarea.tsx
Comment (michelengelen) [reply components/autosize_textarea.tsx:43]: yes ... especially now since we need to support that "overflowing" behavior when the text reaches the button elements on the lower right (send, etc.)

**Comment by M-ZubairAhmed** [comment]
Added QA review, because i assumed this might have potential to break a few e2e. If not feel free to remove it.

**Comment by matthewbirtch** [comment]
@michelengelen did you change the direction of the chevron on the formatting toolbar button as we discussed?

**Comment by michelengelen** [comment]
> @michelengelen did you change the direction of the chevron on the formatting toolbar button as we discussed? I'm not seeing any change there yet

oh no ... i somehow did not push that change! :D

**Comment by matthewbirtch** [comment]
> > @michelengelen did you change the direction of the chevron on the formatting toolbar button as we discussed? I'm not seeing any change there yet
> 
> oh no ... i somehow did not push that change! :D

Ok, I'm good to approve once that's done

**Comment by jgilliam17** [comment]
Thanks @michelengelen 
[First report ](https://mattermost-cypress-report.s3.amazonaws.com/260149-1126e2b-onprem-ent-webapp-pr-11015-mm-ee-test:1126e2b/mochawesome.html) posted with failures related to this PR. Thanks @metanerd for updating and re-running 🤗  see newer [report here](https://mattermost-cypress-report.s3.amazonaws.com/260188-75abd25-onprem-ent-webapp-pr-11015-mm-ee-test:75abd25/mochawesome.html) similar failures
Additionally, I see few issues when testing manually
1. Overlap with file attachments when formatting bar is collapsed 
Screen Shot 2022-11-09 at 11 07 42 AM

2. Multi-attachments are adding horizontal scroll
Screen Shot 2022-11-09 at 11 03 05 AM

3. When toggling formatting bar ON on the rhs, can we scroll rhs up to reveal the bar

https://user-images.githubusercontent.com/52937121/200893952-618b6f10-06b1-402a-a4cd-468867f55ba9.mov


4. Expanding RHS while tiling browser window doesn't expand the formatting bar fully. I think there's room to reveal more buttons 
Screen Shot 2022-11-09 at 11 27 49 AM

**Comment by michelengelen** [comment]
@matthewbirtch please re-review the Chevron direction

@jgilliam17 I tried to fix all PR related failures on the E2E tests. will re-run tests ... could you go through the report and see if there is anything left.

@metanerd 💙 🙇🏼

**Comment by matthewbirtch** [comment]
@michelengelen this may be a different issue, but it looks like we have a lot of extra space when attachments are added to the reply input box (RHS and threads view). I'm going to approve this PR, but can you create a follow-up that fixes this if you don't think it should be included here?
Screenshot 2022-11-16 at 10 12 02 AM

**Review by matthewbirtch — APPROVED** [review]

**Comment by jgilliam17** [comment]
Thanks @michelengelen 
1,2 and 3 are fixed.  Let me know if 4 will also be included in this PR.
Similarly to what Matt mentioned above, RHS textbox expands when formatting bar is toggled on

https://user-images.githubusercontent.com/52937121/202224889-cb3941a6-dcd8-47bb-b2c1-74c11c9510f0.mov

**Comment by jgilliam17** [comment]
@michelengelen Here's the e2e report https://mattermost-cypress-report.s3.amazonaws.com/263260-bc900a7-onprem-ent-webapp-pr-11015-mm-ee-test:bc900a7/mochawesome.html
All of the failures appear PR related, not on master. Can you please take a look? Thanks 🙂

**Comment by michelengelen** [comment]
@jgilliam17 PR #8869 introduced a bug in the reply textbox which breaks a few things here and there. These need to be fixed before I can fix the issues on the E2E tests. Ticket with reference to this PR: https://mattermost.atlassian.net/browse/MM-48618.

cc/ @koox00

**Comment by andrewbrown00** [comment]
This is looking really great everyone! 

I was testing the changes and noticed a delay in growing the height on the input when the text wraps to the second line. It grows the text area slightly, then once it hits a specific character count on the new line it grows again. I didn't see this addressed in the comments above, but let me know if it's already been called out elsewhere.

**Screen recording**


cc @matthewbirtch @jgilliam17 @michelengelen @esethna

**Comment by matthewbirtch** [comment]
> This is looking really great everyone!
> 
> I was testing the changes and noticed a delay in growing the height on the input when the text wraps to the second line. It grows the text area slightly, then once it hits a specific character count on the new line it grows again. I didn't see this addressed in the comments above, but let me know if it's already been called out elsewhere.
> 
> **Screen recording**  [  ](https://user-images.githubusercontent.com/48559658/206756920-6aa62631-224e-4452-9c27-1428d4d0d94c.gif) [ ](https://user-images.githubusercontent.com/48559658/206756920-6aa62631-224e-4452-9c27-1428d4d0d94c.gif)
> 
> cc @matthewbirtch @jgilliam17 @michelengelen @esethna

Hadn't seen this yet. Thanks for calling it out @andrewbrown00.

**Comment by michelengelen** [comment]
@andrewbrown00 @matthewbirtch I fixed that behavior.

**Comment by michelengelen** [comment]
Going through the e2e run it seems some tests are failing, but not from this PR.

- `tests/integration/enterprise/extend_session/email_login_activity_spec.ts` does not work at all
- MM-T1458 works on local
- MM-T904 has a wrong number of rows in this assertion: `cy.get('.admin-console__content .row').should('have.length', 5);` (should be 4 according to the screenshot below)
Screenshot 2022-12-14 at 16 10 57
- MM-T905 should possibly be adjusted as well (maybe the translation changed?)
- MM-T307 does seem if it is incorrect
- `tests/integration/integrations/incoming_webhook/inapp_username_profile_override_spec.js` also seems incorrect (looks like cypress is looking for the wrong class here - should probably be `user-profile-popover__heading`)
- MM-T254 has an incorrect value (so it seems)
- MM-T2212 selector is incorrect: `cy.get(`#postMessageText_${postId}`).find('.mention-link.mention--highlight').should('exist');` (should be `.mention--highlight .mention-link`)
- MM-T453_2 dunno whats wrong here without looking deeper into it
- MM-T473_2 this does seem as if the mention selector is not working as well
- MM-T2368 not sure either whats wrong here
- MM-T2096_2 looks like the `Z` in the timestamp is expected 
- MM-T906 also has an incorrect value for the length (should be 3 according to the screenshot below)
Screenshot 2022-12-14 at 16 26 12
- tests/integration/messaging/message_channel_draw_spec.js - here it seems the Draw plugin is not working as expected

@jgilliam17 or @saturninoabril could you check that please? If it helps I can open a ticket to fix these tests. 

Thanks in advance!

**Comment by jgilliam17** [comment]
Thank you @michelengelen for including detailed list of failures.
I compared [PR 11015 report](https://mattermost-cypress-report.s3.amazonaws.com/275689-1ab1e9a-onprem-ent-webapp-pr-11015-mm-ee-test:1ab1e9a/mochawesome.html) with [master report ](https://mattermost-cypress-report.s3.amazonaws.com/275673-onprem-ent-master-mm-ee-test:test/mochawesome.html) and I think only one test - test case MM-T1458 differs - fails on the PR, but not master and should be addressed. We can disregard other failures as not related.

**Comment by michelengelen** [comment]
@jgilliam17 I fixed the failing test for this. I will restart the E2E tests, so could you check the report and approve latert oday, so we can get this change in asap? Thanks!!! 🙇🏼

**Comment by jgilliam17** [comment]
Thanks @michelengelen 
Looks good to merge.
[E2E report](https://mattermost-cypress-report.s3.amazonaws.com/276172-e8624f0-onprem-ent-webapp-pr-11015-mm-ee-test:e8624f0/mochawesome.html) I see same test failure and will file for follow up/investigation. Otherwise report has no other failures related to the PR.

**Review by jgilliam17 — APPROVED** [review]

**Pull request merged**
